File: backend/app/schemas/plan_semanal.py
from datetime import datetime

from pydantic import BaseModel, Field


class PlanDiaEjercicio(BaseModel):
    nombre_ejercicio: str
    grupo_muscular: str | None = None
    series: int = 3
    repeticiones: str = "10-12"
    descanso_segundos: int = 90
    rir_o_rpe: str | None = None
    notas: str | None = None
    machine_id: str | None = None
    machine_nombre: str | None = None
    machine_foto_url: str | None = None


class PlanDiaBloque(BaseModel):
    tipo: str
    nombre: str
    duracion_minutos: int | None = None
    ejercicios: list[PlanDiaEjercicio] = []


class PlanDia(BaseModel):
    dia_semana: int = Field(..., ge=0, le=6)
    nombre_dia: str
    tipo: str = Field(
        ...,
        description="workout | rest | active_recovery",
        pattern="^(workout|rest|active_recovery)$",
    )
    bloques: list[PlanDiaBloque] = []
    descanso_entre_bloques_minutos: int = 2
    tiempo_total_estimado_minutos: int | None = None
    notas: str | None = None


class PlanSemanalJSON(BaseModel):
    dias: list[PlanDia] = Field(..., min_length=7, max_length=7)
    nota_general: str | None = None


class AIWeeklyPlanRequest(BaseModel):
    objetivo: str = Field(
        ...,
        description="fuerza | hipertrofia | definicion | cardio | funcional | flexibilidad",
    )
    dias_por_semana: int = Field(4, ge=1, le=7)
    duracion_max_minutos: int = Field(75, ge=30, le=150)
    nivel_experiencia: str = Field(
        "intermedio", description="principiante | intermedio | avanzado"
    )
    equipo_disponible: list[str] = [
        "barra",
        "mancuernas",
        "polea",
        "leg_press",
        "smith",
    ]
    lesiones_o_limitaciones: list[str] = []
    notas_adicionales: str | None = None
    maquinas_usuario_ids: list[str] = []
    dias_entreno_seleccionados: list[str] = []
    preferencias_equipamiento: list[str] = []
    porcentaje_maquinas_guiadas: int | None = Field(
        None, ge=0, le=100, description="Porcentaje deseado de máquinas guiadas"
    )
    porcentaje_peso_libre: int | None = Field(
        None, ge=0, le=100, description="Porcentaje deseado de peso libre y mancuernas"
    )
    min_ejercicios_por_sesion: int = Field(
        4, ge=2, le=12, description="Mínimo de ejercicios en el bloque principal por sesión"
    )


class PlanSemanalCreate(BaseModel):
    nombre: str
    objetivo: str
    nivel: str = "intermedio"
    duracion_max_minutos: int = 75
    dias_entreno_objetivo: int = 4
    equipo_disponible: list[str] = ["barra", "mancuernas", "polea"]
    lesiones_o_limitaciones: list[str] = []
    plan_json: dict
    metadata_ia: dict | None = None


class PlanSemanalResponse(BaseModel):
    id: str
    usuario_id: str
    nombre: str
    objetivo: str
    nivel: str
    duracion_max_minutos: int
    dias_entreno_objetivo: int
    equipo_disponible: list[str]
    lesiones_o_limitaciones: list[str] | None
    plan_json: dict
    metadata_ia: dict | None
    activo: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PlanSemanalListResponse(BaseModel):
    id: str
    nombre: str
    objetivo: str
    nivel: str
    activo: bool
    created_at: datetime

    class Config:
        from_attributes = True


class AIWeeklyPlanResponse(BaseModel):
    ok: bool = True
    plan_guardado: PlanSemanalResponse
    proveedor: str
    modelo: str
    modo: str


class AIModifyPlanRequest(BaseModel):
    plan_id: str
    instrucciones: str

File: flutter_app/lib/features/weekly_plan/domain/plan_semanal.dart
import '../../../core/app_config.dart';
import '../../../core/env.dart';

class PlanDiaEjercicio {
  final String nombreEjercicio;
  final String? grupoMuscular;
  final int series;
  final String repeticiones;
  final int descansoSegundos;
  final String? rirORpe;
  final String? notas;
  final String? machineId;
  final String? machineNombre;
  final String? machineFotoUrl;

  String? get imageUrl {
    if (machineFotoUrl == null) return null;
    final activeBaseUrl = AppConfig.baseUrl ?? Env.apiBaseUrl;
    final baseUrl = activeBaseUrl.replaceAll('/api/v1', '');
    final cleanPath = machineFotoUrl!.replaceAll('backend/', '').replaceAll('backend/storage/', 'storage/');
    final path = cleanPath.startsWith('/') ? cleanPath : '/$cleanPath';
    if (path.contains('/api/v1/')) {
      return '$baseUrl$path';
    } else {
      final formattedPath = path.replaceAll('/storage/', '/api/v1/storage/');
      return '$baseUrl$formattedPath';
    }
  }

  PlanDiaEjercicio({
    required this.nombreEjercicio,
    this.grupoMuscular,
    this.series = 3,
    this.repeticiones = '10-12',
    this.descansoSegundos = 90,
    this.rirORpe,
    this.notas,
    this.machineId,
    this.machineNombre,
    this.machineFotoUrl,
  });

  factory PlanDiaEjercicio.fromJson(Map<String, dynamic> json) {
    return PlanDiaEjercicio(
      nombreEjercicio: json['nombre_ejercicio'] as String? ?? '',
      grupoMuscular: json['grupo_muscular'] as String?,
      series: json['series'] as int? ?? 3,
      repeticiones: json['repeticiones'] as String? ?? '10-12',
      descansoSegundos: json['descanso_segundos'] as int? ?? 90,
      rirORpe: json['rir_o_rpe'] as String?,
      notas: json['notas'] as String?,
      machineId: json['machine_id'] as String?,
      machineNombre: json['machine_nombre'] as String?,
      machineFotoUrl: json['machine_foto_url'] as String?,
    );
  }
}

class PlanDiaBloque {
  final String tipo;
  final String nombre;
  final int? duracionMinutos;
  final List<PlanDiaEjercicio> ejercicios;

  PlanDiaBloque({
    required this.tipo,
    required this.nombre,
    this.duracionMinutos,
    this.ejercicios = const [],
  });

  factory PlanDiaBloque.fromJson(Map<String, dynamic> json) {
    return PlanDiaBloque(
      tipo: json['tipo'] as String? ?? '',
      nombre: json['nombre'] as String? ?? '',
      duracionMinutos: json['duracion_minutos'] as int?,
      ejercicios: (json['ejercicios'] as List<dynamic>?)
              ?.map((e) => PlanDiaEjercicio.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class PlanDia {
  final int diaSemana;
  final String nombreDia;
  final String tipo;
  final List<PlanDiaBloque> bloques;
  final int descansoEntreBloquesMinutos;
  final int? tiempoTotalEstimadoMinutos;
  final String? notas;

  PlanDia({
    required this.diaSemana,
    required this.nombreDia,
    required this.tipo,
    this.bloques = const [],
    this.descansoEntreBloquesMinutos = 2,
    this.tiempoTotalEstimadoMinutos,
    this.notas,
  });

  bool get isWorkout => tipo == 'workout';
  bool get isRest => tipo == 'rest';
  bool get isActiveRecovery => tipo == 'active_recovery';

  factory PlanDia.fromJson(Map<String, dynamic> json) {
    return PlanDia(
      diaSemana: json['dia_semana'] as int? ?? 0,
      nombreDia: json['nombre_dia'] as String? ?? '',
      tipo: json['tipo'] as String? ?? 'rest',
      bloques: (json['bloques'] as List<dynamic>?)
              ?.map((e) => PlanDiaBloque.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      descansoEntreBloquesMinutos: json['descanso_entre_bloques_minutos'] as int? ?? 2,
      tiempoTotalEstimadoMinutos: json['tiempo_total_estimado_minutos'] as int?,
      notas: json['notas'] as String?,
    );
  }
}

class PlanSemanal {
  final String id;
  final String usuarioId;
  final String nombre;
  final String objetivo;
  final String nivel;
  final int duracionMaxMinutos;
  final int diasEntrenoObjetivo;
  final List<String> equipoDisponible;
  final List<String>? lesionesOLimitaciones;
  final PlanSemanalJSON planJson;
  final Map<String, dynamic>? metadataIa;
  final bool activo;
  final DateTime createdAt;
  final DateTime updatedAt;

  PlanSemanal({
    required this.id,
    required this.usuarioId,
    required this.nombre,
    required this.objetivo,
    required this.nivel,
    this.duracionMaxMinutos = 75,
    this.diasEntrenoObjetivo = 4,
    this.equipoDisponible = const [],
    this.lesionesOLimitaciones,
    required this.planJson,
    this.metadataIa,
    this.activo = true,
    required this.createdAt,
    required this.updatedAt,
  });

  PlanSemanal copyWith({
    bool? activo,
  }) {
    return PlanSemanal(
      id: id,
      usuarioId: usuarioId,
      nombre: nombre,
      objetivo: objetivo,
      nivel: nivel,
      duracionMaxMinutos: duracionMaxMinutos,
      diasEntrenoObjetivo: diasEntrenoObjetivo,
      equipoDisponible: equipoDisponible,
      lesionesOLimitaciones: lesionesOLimitaciones,
      planJson: planJson,
      metadataIa: metadataIa,
      activo: activo ?? this.activo,
      createdAt: createdAt,
      updatedAt: updatedAt,
    );
  }

  factory PlanSemanal.fromJson(Map<String, dynamic> json) {
    return PlanSemanal(
      id: json['id'] as String,
      usuarioId: json['usuario_id'] as String,
      nombre: json['nombre'] as String,
      objetivo: json['objetivo'] as String,
      nivel: json['nivel'] as String? ?? 'intermedio',
      duracionMaxMinutos: json['duracion_max_minutos'] as int? ?? 75,
      diasEntrenoObjetivo: json['dias_entreno_objetivo'] as int? ?? 4,
      equipoDisponible: (json['equipo_disponible'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      lesionesOLimitaciones: (json['lesiones_o_limitaciones'] as List<dynamic>?)
          ?.map((e) => e as String)
          .toList(),
      planJson: PlanSemanalJSON.fromJson(json['plan_json'] as Map<String, dynamic>),
      metadataIa: json['metadata_ia'] as Map<String, dynamic>?,
      activo: json['activo'] as bool? ?? true,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

class PlanSemanalJSON {
  final List<PlanDia> dias;
  final String? notaGeneral;

  PlanSemanalJSON({required this.dias, this.notaGeneral});

  factory PlanSemanalJSON.fromJson(Map<String, dynamic> json) {
    return PlanSemanalJSON(
      dias: (json['dias'] as List<dynamic>?)
              ?.map((e) => PlanDia.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      notaGeneral: json['nota_general'] as String?,
    );
  }
}